src/SArray.jl

```diff
@@ -50,6 +50,34 @@ end
     end
 end
 
+@generated function (::Type{SArray{S, T}})(gen::Base.Generator) where {S <: Tuple, T}
+    stmts = [:(Base.@_inline_meta)]
+    args = []
+    iter = :(iterate(gen))
+    for i in CartesianIndices(size_to_tuple(S))
+        el = Symbol(:el, i)
+        push!(stmts, :(($el,st) = $iter))
```

Comment:
The error message may be a bit less cryptic if we use

```suggestion
        push!(stmts, :(($el,st) = something($iter)))
```

Or maybe something like `throw_if_nothing($iter, $i)` with appropriate index-aware error message would be even better (at the cost of the compiler doing more works).

Reply:
Yes, I like the `throw_if_nothing` approach.